test/terraform_ssh_example_test.go

```diff
@@ -211,3 +220,208 @@ func testSCPToPublicHost(t *testing.T, terraformOptions *terraform.Options, keyP
 		return "", nil
 	})
 }
+
+func testSSHAgentToPublicHost(t *testing.T, terraformOptions *terraform.Options, keyPair *aws.Ec2Keypair) {
+	// Run `terraform output` to get the value of an output variable
+	publicInstanceIP := terraform.Output(t, terraformOptions, "public_instance_ip")
+
+	// We're going to try to SSH to the instance IP, using the Key Pair we created earlier. Instead of
+	// directly using the SSH key in the SSH connection, we're going to rely on an existing SSH agent that we
+	// programatically emulate within this test. We're going to use the user "ubuntu" as we know the Instance
+	// is running an Ubuntu AMI that has such a user
+	publicHost := ssh.Host{
+		Hostname:    publicInstanceIP,
+		SshUserName: "ubuntu",
+		SshAgent:    true,
+	}
+
+	// It can take a minute or so for the Instance to boot up, so retry a few times
+	maxRetries := 30
+	timeBetweenRetries := 5 * time.Second
+	description := fmt.Sprintf("SSH with Agent to public host %s", publicInstanceIP)
+
+	// Run a simple echo command on the server
+	expectedText := "Hello, World"
+	command := fmt.Sprintf("echo -n '%s'", expectedText)
+
+	// Instantiate a temporary SSH agent
+	rHex, err := randomHex(20)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+	socketFile := "/tmp/ssh-agent-" + rHex + ".sock"
+	os.Setenv("SSH_AUTH_SOCK", socketFile)
+	sshAgent := NewSSHAgent(socketFile)
+	defer sshAgent.Stop()
+
+	// Create SSH key for the agent using the existing AWS SSH key pair
+	block, _ := pem.Decode([]byte(keyPair.KeyPair.PrivateKey))
+	pkey, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+	key := agent.AddedKey{PrivateKey: pkey}
+
+	// Verify that we can SSH to the Instance and run commands
+	retry.DoWithRetry(t, description, maxRetries, timeBetweenRetries, func() (string, error) {
+
+		// Add SSH key to the agent
+		keys, err := sshAgent.agent.List()
+		if len(keys) == 0 {
+			sshAgent.agent.Add(key)
+		}
```

Comment:
Shouldn't you add the key outside of the retry loop?

Reply:
Adding the key to the SSH agent may fail the first time if the service is not yet ready, which happens pretty consistently. Do you agree in adding a 1 or .5 second sleep to avoid this first error would be an acceptable solution considering this is a unit-test?

Reply:
Ah, I see. I think retrying in that case is a more robust solution than a sleep. Just make sure to:

1. Add a comment with the explanation you just wrote
1. Don't ignore the `err` returned by `sshAgent.agent.List()`!
1. Does `sshAgent.agent.Add()` return an error? If so, catch that and force a retry after a brief sleep.